rfcs/20230622-quantized-reduction.md

```diff
@@ -0,0 +1,521 @@
+            -> tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>
+        "stablehlo.return"(%add)
+            : (tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>) -> ()
+    }) {
+        dimensions = dense<...> : tensor<1xi64>
+    } : (tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>,
+         tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>)
+    -> tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>
+```
+
+should be interpreted as
+
+```mlir
+%result = "stablehlo.reduce"(%input, %init_value) ({
+    ^input_conversion(
+            %input: tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>):
+        "stablehlo.return"(%input)
+            : (tensor<!quant.uniform<ui8:f32, input_scale:0>>) -> ()
+
+    }, {
+    ^reduce_computation(
+            %lhs: tensor<!quant.uniform<ui8:f32, input_scale:0>>,
+            %rhs: tensor<!quant.uniform<ui8:f32, input_scale:0>>):
+        %add = "stablehlo.add"(%lhs, %rhs)
+            : (tensor<!quant.uniform<ui8:f32, input_scale:0>>,
+               tensor<!quant.uniform<ui8:f32, input_scale:0>>)
+            -> tensor<!quant.uniform<ui8:f32, input_scale:0>>
+        "stablehlo.return"(%add)
+            : (tensor<!quant.uniform<ui8:f32, input_scale:0>>) -> ()
+    }, {
+    ^output_conversion(
+            %intermediate_result: tensor<!quant.uniform<ui8:f32, input_scale:0>>):
+        "stablehlo.return"(%intermediate_result)
+            : (tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>) -> ()
+    }) {
+        dimensions = dense<...> : tensor<1xi64>
+    } : (tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>,
+         tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>)
+    -> tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>
+```
+
+Note that with default values, the  input/result type of `reduce` op matches
+with the argument or the result type of the `reduce_computation`, including the
+quantization parameters.
+
+It is important to enforce the following aspect in the implementations: both
+conversion functions must be provided or both must be missing. If this is not
+followed, and there are two functions provided in the IR, it would be difficult
+to determine which function is intended for which purpose.
+
+The existing pretty printing is currently producing the following output
+`stablehlo.reduce(%input init: %init_value) applies stablehlo.add across
+dimensions = [1] : (tensor<1x6xi64>, tensor<i64>) -> tensor<1xi64>`. IMO,
+modifying the above format, with the default conversion function, will create
+clutter. My proposal here is to follow the existing pretty printing when the
+conversion functions are "not provided". In the event, the conversion functions
+are explicitly provided, then the pretty printers will fall back to default
+generic printing,
+**even if the explicitly provided conversion functions are identity function**:
+To avoid identification of identity functions which could be tricky in general.
```

Comment:
I think we can limit identity to:

```
^block(%arg: TYPE):
    return %arg : TYPE
```

and ignore other semantic identity function (e.g. `%arg + 0`).

Reply:
SG!